crates/chain/src/keychain/txout_index.rs

```diff
@@ -3,16 +3,95 @@ use crate::{
     indexed_tx_graph::Indexer,
     miniscript::{Descriptor, DescriptorPublicKey},
     spk_iter::BIP32_MAX_INDEX,
-    SpkIterator, SpkTxOutIndex,
+    DescriptorExt, SpkIterator, SpkTxOutIndex,
+};
+use bitcoin::{
+    hashes::{hash_newtype, sha256},
+    OutPoint, Script, Transaction, TxOut, Txid,
 };
-use bitcoin::{OutPoint, Script, Transaction, TxOut, Txid};
 use core::{
     fmt::Debug,
     ops::{Bound, RangeBounds},
 };
 
 use crate::Append;
 
+hash_newtype! {
+    /// Represents the ID of a descriptor, defined as the sha256 hash of
+    /// the descriptor string, checksum included.
+    pub struct DescriptorId(pub sha256::Hash);
+}
+
+/// Represents updates to the derivation index of a [`KeychainTxOutIndex`].
+/// It maps each keychain `K` to a descriptor and its last revealed index.
+///
+/// It can be applied to [`KeychainTxOutIndex`] with [`apply_changeset`]. [`ChangeSet] are
+/// monotone in that they will never decrease the revealed derivation index.
+///
+/// [`KeychainTxOutIndex`]: crate::keychain::KeychainTxOutIndex
+/// [`apply_changeset`]: crate::keychain::KeychainTxOutIndex::apply_changeset
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "K: Ord + serde::Deserialize<'de>",
+            serialize = "K: Ord + serde::Serialize"
+        )
+    )
+)]
+#[must_use]
+pub struct ChangeSet<K> {
+    /// Contains the keychains that have been added and their respective descriptor
+    pub keychains_added: BTreeMap<K, Descriptor<DescriptorPublicKey>>,
+    /// Contains for each descriptor_id the last revealed index of derivation
+    pub last_revealed: BTreeMap<DescriptorId, u32>,
+}
+
+impl<K: Ord> Append for ChangeSet<K> {
+    /// Append another [`ChangeSet`] into self.
+    ///
+    /// For each keychain in `keychains_added` in the given [`ChangeSet`]:
+    /// If the keychain already exist with a different descriptor, we overwrite the old descriptor.
+    ///
+    /// For each `last_revealed` in the given [`ChangeSet`]:
+    /// If the keychain already exists, increase the index when the other's index > self's index.
+    fn append(&mut self, mut other: Self) {
+        for (keychain, descriptor) in &mut self.keychains_added {
+            if let Some(other_descriptor) = other.keychains_added.remove(keychain) {
+                *descriptor = other_descriptor;
+            }
+        }
+
+        for (descriptor_id, index) in &mut self.last_revealed {
+            if let Some(other_index) = other.last_revealed.remove(descriptor_id) {
+                *index = other_index.max(*index);
+            }
+        }
+
```

Comment:
For updating `last_revealed`, I think it's better to just loop through `other.last_revealed` and get rid of the call to `self.last_revealed.extend`.

Reply:
Can you explain why?

Reply:
Because `other` is changing `self`. Looping through changes and applying it to `self` is the more straightforward logic.

Reply:
I agree that @evanlinjin 's approach is easier to read and looks to be more efficient with only one loop through `other.last_revealed`. The way I read it is: 
1. extend my keychains_added with other's keychains_added
2. update my last_revealed with just the changes in other's last_revealed

Reply:
BTW, based on recent comments step 1. above will be more complicated since it needs to check that `K` isn't pointing to a different descriptor and `Descriptor` isn't assigned to a different `K`; in either case don't add the new keychain right?